extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js

```diff
@@ -105,45 +109,213 @@ export class AmpStoryShoppingAttachment extends AMP.BaseElement {
-
-    if (isOpen && isOnActivePage) {
-      this.populatePlp_();
+  checkClearActiveProductData_(isOpen, shoppingData) {
+    const {activeProductData} = shoppingData;
+    if (!isOpen && activeProductData && this.isOnActivePage_()) {
+      this.storeService_.dispatch(Action.ADD_SHOPPING_DATA, {
+        'activeProductData': null,
+      });
     }
   }
 
   /**
-   * Renders a list of the products on the page, knows as a "Product Listing Page" or PLP.
+   * On attachment state or shopping data state update, check if on active page and update template.
+   * @param {boolean} isOpen
+   * @param {!Array<!ShoppingConfigDataDef>} shoppingData
    * @private
    */
-  populatePlp_() {
-    if (this.plpContainer_.querySelector('.i-amphtml-amp-story-shopping-plp')) {
+  updateTemplate_(isOpen, shoppingData) {
+    if (!isOpen || !this.isOnActivePage_()) {
       return;
     }
-    const shoppingData = this.storeService_.get(StateProperty.SHOPPING_DATA);
     const shoppingDataForPage = this.shoppingTags_.map(
       (shoppingTag) => shoppingData[shoppingTag.getAttribute('data-product-id')]
     );
 
-    const plp = this.renderPlpTemplate_(shoppingDataForPage);
-    this.mutateElement(() => {
-      this.plpContainer_.appendChild(plp);
+    // If there is only one product on the page, feature it.
+    // Otherwise feature the active product.
+    const singleProductOnPage =
+      shoppingDataForPage.length === 1 && shoppingDataForPage[0];
+    const featuredProduct =
+      shoppingData.activeProductData || singleProductOnPage;
+
+    // templateId string used to key already built templates.
+    const templateId = `pdp-${featuredProduct?.productId || 'plp'}`;
+
+    // Iterate over built templates and set active attribute.
+    // If a template is already built, return early.
+    let templateAlreadyBuilt = false;
+    Object.values(this.builtTemplates_).forEach((template) => {
+      if (template === this.builtTemplates_[templateId]) {
+        templateAlreadyBuilt = true;
+        this.mutateElement(() => {
+          template.setAttribute('active', true);
+          this.resetCarouselScroll_(template);
+        });
+      } else {
+        this.mutateElement(() => template.removeAttribute('active'));
+      }
+    });
+
+    // Early return if template is already built.
+    if (templateAlreadyBuilt) {
+      return;
+    }
+
+    // If not already built, build template and assign it to builtTemplates_ by templateId.
+    this.builtTemplates_[templateId] = (
+      <div class="i-amphtml-amp-story-shopping" active>
+        {featuredProduct && this.renderPdpTemplate_(featuredProduct)}
+        {shoppingDataForPage.length > 1 &&
+          this.renderPlpTemplate_(
+            shoppingDataForPage.filter((item) => item !== featuredProduct)
+          )}
+      </div>
+    );
+
+    this.mutateElement(() =>
+      this.templateContainer_.appendChild(this.builtTemplates_[templateId])
+    );
```

Comment:
This piece of code is mostly great but has a slight design issue, it creates two code paths depending on if the template is created or re-used.
Down the line this creates more complexity, e.g. you only resetCarouselScroll when re-using the template, not creating it. If you scroll on a PLP and click on a product, the scroll reset behavior would already be different.
You also shouldn't set `active` directly on the template as this is an option, if you wanted to prerender the templates that wouldn't work out of the box. 

You could extract the template creation into a separate method, like

```
// pseudocode
getTemplate(templateId, featuredProduct, shoppingDataForPage) {
  return this.builtTemplates_[templateId] || (jsx);
}
```

And then in your updateTemplate:

```
// Pseudocode
updateTemplate(...) {
  ...

  forEach(template => removeAttr('active'));
  const template = getTemplate(...);
  // append template if !template.isConnected, set template active, resetScroll
}
```